### ISSUE TITLE
use-sound hook is broken after installing it on a typescript project.

### ISSUE THREAD

**Issue by mnsdojo** [joshwcomeau/use-sound#135]
```js
Could not find a declaration file for module 'use-sound'. 'c:/Users/manish/Documents/auth-fire/node_modules/use-sound/dist/use-sound.esm.js' implicitly has an 'any' type.
  There are types at 'c:/Users/manish/Documents/auth-fire/node_modules/use-sound/dist/index.d.ts', but this result could not be resolved when respecting package.json "exports". The 'use-sound' library may need to update its package.json or typings.ts(7016)
Could not find a declaration file for module 'use-sound'. 'c:/Users/manish/Documents/auth-fire/node_modules/use-```

**Comment by ladvoc** [comment]
I am receiving the same error:
- use-sound version: v4.0
- React version: 18.2.0
- Vite: 4.4.5
- Typescript: 5.1.3
- Platform: macOS

**Comment by hemantwasthere** [comment]
Does anyone know how can we temporary solve this error like jumping on the previous versions and use that?

**Comment by SebassNoob** [comment]
Same here. Temporarily fixed by adding ``// @ts-ignore`` on top of the import, but TypeScript will not be able to provide any type safety for the hook itself. Personally going to hold off from using this package for now.

**Comment by ProgrammerSteve** [comment]
What I did was make a use-sound.d.ts file
Then in the tsconfig.json file add    
"typeRoots": ["./use-sound.d.ts"]
inside the compilerOptions


The use-sound.d.ts file looked like this
```
declare module 'use-sound' {
  import { Howl, Howler } from 'howler';

  interface HookOptions {
    volume?: number;
    playbackRate?: number;
    interrupt?: boolean;
    soundEnabled?: boolean;
    sprite?: { [key: string]: [number, number] };
    // You can add more properties as needed based on your use case
  }

  interface ExposedData {
    volume: number;
    playbackRate: number;
    interrupt: boolean;
    soundEnabled: boolean;
    sprite: { [key: string]: [number, number] };
    // You can add more properties as needed based on your use case
  }

  type PlayFunction = (options?: PlayOptions) => void;

  interface PlayOptions {
    id?: string;
    forceSoundEnabled?: boolean;
    playbackRate?: number;
    // You can add more properties as needed based on your use case
  }

  interface PlayExposedData extends ExposedData {
    stop: (id?: string) => void;
    pause: (id?: string) => void;
    duration: number | null;
    sound: Howl | null;
  }

  type UseSoundTuple = [PlayFunction, PlayExposedData];

  const useSound: (src: string, options?: HookOptions) => UseSoundTuple;

  export default useSound;
}
```

**Comment by Pushplaybang** [comment]
same issue with latest vite/ react/ts. added `// @ts-igmore` above the import.

**Comment by akhil0001** [comment]
This helped resolve my issue. 

1. Create a folder called `generated-types` at the same level of tsconfig.json
2. Now, copy the `index.d.ts` and `types.d.ts` files from dist folder of `@node-modules/use-sound` and paste them in the above folder
3. Head over to `tsconfig.json` and add `paths` to it in the below format inside `compilerOptions`
```
"paths": {
      "use-sound": ["./generated-types/"]
    }
```

This might or might not be the exact correct answer, but this is helpful than `// @ts-ignore`

**Comment by hazadus** [comment]
Same stuff, using Next.js 13 and Typescript.

**Comment by dhino12** [comment]
> This helped resolve my issue.
> 
> 1. Create a folder called `generated-types` at the same level of tsconfig.json
> 2. Now, copy the `index.d.ts` and `types.d.ts` files from dist folder of `@node-modules/use-sound` and paste them in the above folder
> 3. Head over to `tsconfig.json` and add `paths` to it in the below format inside `compilerOptions`
> 
> ```
> "paths": {
>       "use-sound": ["./generated-types/"]
>     }
> ```
> 
> This might or might not be the exact correct answer, but this is helpful than `// @ts-ignore`
> 
> _Note: The folder `generated-types` can be named anything and the same name should be reflecting inside the `use-sound` of the `paths` key in tsconfig.json_

it works for me on Next JS 13, thanks

**Comment by kam-st** [comment]
@dhino12 solution works. But why does it not pickup types, where they are in nodemodules?

**Comment by manseerat21** [comment]
@dhino12 thank you.

**Comment by manseerat21** [comment]
thank you @akhil0001 @dhino12

**Comment by jeffclaybrook** [comment]
@akhil0001 you da man thank you!

**Comment by ackDeckard** [comment]
@akhil0001 solution is working even on Next.js 14. Thank you!

**Comment by bhatvikrant** [comment]
If someone is still getting TS errors while using the useSound hook, then try installing @types/howler along with the above mentioned solution.

```
bun install --dev @types/howler 
```

**Comment by memark** [comment]
> If someone is still getting TS errors while using the useSound hook, then try installing @types/howler along with the above mentioned solution.
> 
> ```
> bun install --dev @types/howler 
> ```

This is the way.

**Comment by kamaci141** [comment]
@akhil0001 thank you

**Comment by dulrich2222** [comment]
Using @akhil0001 solution idea, this worked for me and I'm using Next JS 14.

In the tsconfig.json file, add "use-sound": ["./node_modules/use-sound/"] to paths like this

```
"paths": {
  "@/*": ["./*"],
  "use-sound": ["./node_modules/use-sound/"]
}
```

**Comment by joshwcomeau** [comment]
Sorry for the trouble y'all — this would be fixed if we switched to ESM-only, but then it would break for anyone using CommonJS. Maybe we'll do that in a future version, but I don't want to break the library for anyone, or maintain two separate versions.

I tested @akhil0001’s fix and it worked for me! I don't expect the type interface to change, so it should be safe to copy these files. And if it ever does change, it'd be a major version bump which would hopefully remove the need to jump through these hoops.
